Adding Frontend/API testing functionality to pipeline/stacks.

DillonRanwala opened the pull request.

DillonRanwala committed:
feat(cdk): add ttl attribute to all existing DynamoDB tables and update
visit/register lambdas to use ttl attribute.

Problem:

Test items from pipelines or canary testing will create dummy data that
we would like to remove after a certain time period. TTL settings on
Dynamo tables allow us to automatically remove any items with a ttl
attribute. Also, we need to update our visit/register lambdas to make
sure they pass the ttl attribute from the API call to our tables.

Solution:

Enabled the ttl attribute with a "ttl_expiration" attribute for all 4
tables. Updated the visit and register lambdas to extract a
ttl_expiration value from any incoming requests (if it exists) and
insert it into the correct table. Also this change will ensure both the
original and updated tables both have the same status code before
returning a 200 from any API call.

Testing:

Created ttl Dynamo tables and test items in a dev environment. Updated
dev lambdas and verified that ttl worked and deleted only the correct
items.
Issue:
N/A, no open GitHub Issue.

DillonRanwala committed:
test(cdk): add curl tests to CodePipeline for Beta and Prod urls

Problem:

In order to make sure any new code changes didn't break our front-end,
we can curl the Cloudfront urls to ensure we get a response.

Solution:

Added a Shellstep to both the Beta and Prod stages in Pipeline.py that
run a simple curl command on the Cloudfront endpoints hosting the login
system.

Testing:

This testing shellstep was deployed in a dev CodePipeline and ran
successfully when the dev endpoint was running.

Issue:
N/A, No open issue.

DillonRanwala committed:
test(cdk/visit): adds API testing lambda and python script for later
integration with pipeline and canary testing.

Problem:

In order to increase the amount of automated testing, we can verify our
API Gateway, Lambda, and Cloudfront services are functional with API
test scripts. We need to be able to automatically run these tests
whenever a new change is pushed and run tests continuously as a canary.

Solution:

This change adds a "TestAPILambda" function to the visit stack and
creates a testing_script.py. These files run the same tests, but the
lambda to be used in a canary test (Eventbridge trigger that can run the
lambda on some set timeframe) while the python script will be called in
a CodePipeline step. The tests will curl the proper frontend, call the
visit/register endpoints with dummy data (CANARY_TEST and PIPELINE_TEST
prefixes for the usernames), and verify that each endpoint is
functional. Note that this commit only adds the code to the
repo/Cloudformation. Later commits/changes will set up the tests to
actually run.

Testing:

The lambda and script were added to Dev visit stacks and CodePipelines
and displayed the correct behavior for situations where it should both
succeed and fail.

Issue:
N/A, No open issue.

DillonRanwala committed:
test(cdk): adds step in pipeline to call testing_script.py in Beta and
Prod

Problem:

Need to integrate the API testing functionality into our CodePipeline.
If added as a step in our pipeline's stages, the tests will run
automatically every time any code is pushed to Github.

Solution:

Adds Shellsteps for testing API endpoints to both Beta and Prod stages
in Pipeline.py. Uses a python command to call testing_script.py in the
pipeline. Also uses the codestar connection as a input source for each
step.

Testing:

Shellsteps were added to Dev pipelines and ran successfully on each code
update.

Issue:
N/A, No open issue.

dillongorsen commented on cdk/database/__init__.py:
```diff
@@ -71,7 +71,8 @@ def dynamodb_old_table(self):
                                                      type=aws_dynamodb.AttributeType.STRING),
-                                                 billing_mode=aws_dynamodb.BillingMode.PAY_PER_REQUEST)
+                                                 billing_mode=aws_dynamodb.BillingMode.PAY_PER_REQUEST,
+                                                 time_to_live_attribute="ttl_expiration")
```
probably makes more sense to call this something like "last_updated" but not a huge deal

dillongorsen commented on .github/workflows/pytest.yml:
```diff
@@ -29,3 +29,4 @@ jobs:
                   USERS_TABLE_NAME: users
                   VISITS_TABLE_NAME: visits
                   DOMAIN_NAME: https://visit.cumaker.space
+                  ENV: Beta
```
out of curiosity, why ENV: Beta with the prod domain name?

DillonRanwala replied:
At the time I was just trying to set a variable that would pass the pytest command, and I didn't think there would be a functional difference between "Beta" and "Prod." However, looking at it now, it probably makes sense to just use Prod.

dillongorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -64,6 +64,12 @@ def __init__(self, original_table, users_table, dynamodbclient):
 
+        # get ttl_expiration from test users if they exist
+        try:
+            ttl_expiration = {"N":str(user_info['ttl_expiration'])}
+        except:
+            ttl_expiration = {"S":""}
```
The mixed attribute type here makes me a bit nervous.  What was the rationale?

DillonRanwala replied:
The ttl functionality only triggers if the dynamo table attribute is a Number type. All the testing elements should have this ttl_expiration attribute (so they won't raise the exception) and be a Number type in AWS. I wanted to make sure no non-testing items would have ttl triggered by accident, so I just set the ttl_expiration to an empty String type. If it was an empty Number type, I wasn't sure if it would cause any deletions, so to be safe I just changed it to a String type.

dillongorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -77,7 +83,8 @@ def add_user_info(self, user_info):
                 'Major': ', '.join(sorted(user_info.get('Major', []))),
-                'Minor': ', '.join(sorted(user_info.get('Minor', [])))
+                'Minor': ', '.join(sorted(user_info.get('Minor', []))),
+                'ttl_expiration':ttl_expiration,
```
Nitpick but this should adhere to the same style as the rest of the table. TtlExpriation

DillonRanwala replied:
I think I can change this one to keep the same formatting.

ophilli replied:
In your latest commits it looks like you didn’t take this suggestion. Is that intentional?

DillonRanwala replied:
For using the "last_updated" format instead of something like "LastUpdated", I think it is better to keep the TTL attribute names uniform across all DynamoDB tables. While it could be helpful for this one table to have uniform formatting, I think keeping every ttl attribute name the same (as "last_updated") would probably be simpler if future teams need to interact with the items.

DillonRanwala committed:
refactor(cdk): adjust formatting of ttl usage and other formatting
changes/

Problem:

Needed to update formatting of ttl usage in the dynamo and lambda code.

Solution:

Changes ttl_expiration attribute name to last_updated. Keeps this
attribute format uniform in testing code (only uses a Number type). Also
changes github workflow pytest env variable to Prod from Beta.

Testing:

Issue:
N/A, no open GitHub Issue.

ophilli commented:
😔 This branch cannot be rebased due to conflicts. Perhaps you need to rebase & force push onto your branch? Or perhaps you need to eat it and make a merge commit.

DillonRanwala commented:
I tried rebasing with a force push, but that didn't seem to do anything. On my end, I'm not seeing any merge conflicts, but I can still try to make a merge commit.

DillonRanwala commented:
> I tried rebasing with a force push, but that didn't seem to do anything. On my end, I'm not seeing any merge conflicts, but I can still try to make a merge commit.

Update, I didn't rebase properly the first time. I dropped a commit which I think might have been the issue, and I'm not seeing any merge conflicts on my end.

DillonRanwala commented:
Notes on testing in Beta before approving this change to Prod. 

- CodePipeline successfully ran the ShellStep tests for curling the front end and testing the APIs with dummy data. 
- Also, the TestAPILambda was successfully created and succeeded when tested (via the Lambda service test option). 
- The Beta websites were working with the Cloudfront distribution. 
- Normal visit requests were handled correctly by the API gateway and lambdas while being put in the DynamoDB tables. 
- The Beta tables were changed to have a TTL attribute, and the test items from the pipeline and testing canary were properly deleted automatically from all tables (by using the last_updated field). 
- New and existing non-test items were not affected by the new TTL setting. 

Therefore, I feel confident in pushing this change to Production.
